tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -2572,6 +2572,108 @@ void SwiftLangSupport::findRelatedIdentifiersInFile(
                           llvm::vfs::getRealFileSystem());
 }
 
+//===----------------------------------------------------------------------===//
+// SwiftLangSupport::findInactiveRangesInFile
+//===----------------------------------------------------------------------===//
+
+namespace {
+class IfConfigScanner : public SourceEntityWalker {
+  SmallVectorImpl<IfConfigInfo> &Infos;
+  SourceManager &SourceMgr;
+  unsigned BufferID = -1;
+  bool Cancelled = false;
+
+public:
+  explicit IfConfigScanner(SourceFile &SrcFile, unsigned BufferID,
+                           SmallVectorImpl<IfConfigInfo> &Infos)
+      : Infos(Infos), SourceMgr(SrcFile.getASTContext().SourceMgr),
+        BufferID(BufferID) {}
+
+private:
+  bool walkToDeclPre(Decl *D, CharSourceRange Range) override {
+    if (Cancelled)
+      return false;
+
+    if (auto *IfDecl = dyn_cast<IfConfigDecl>(D)) {
+      for (auto &Clause : IfDecl->getClauses()) {
+        unsigned Offset = SourceMgr.getLocOffsetInBuffer(Clause.Loc, BufferID);
+        IfConfigInfo Info;
+        Info.Offset = Offset;
+        Info.IsActive = Clause.isActive;
+        Infos.emplace_back(Info);
+      }
+    }
+
+    return true;
+  }
+};
+
+} // end anonymous namespace
+
+void SwiftLangSupport::findInactiveRangesInFile(
+    StringRef InputFile, bool CancelOnSubsequentRequest,
+    ArrayRef<const char *> Args, SourceKitCancellationToken CancellationToken,
+    std::function<void(const RequestResult<InactiveRangesInfo> &)> Receiver) {
+
+  std::string Error;
+  SwiftInvocationRef Invok =
+      ASTMgr->getTypecheckInvocation(Args, InputFile, Error);
+  if (!Invok) {
+    LOG_WARN_FUNC("failed to create an ASTInvocation: " << Error);
+    Receiver(RequestResult<InactiveRangesInfo>::fromError(Error));
+    return;
+  }
+
+  class IfConfigConsumer : public SwiftASTConsumer {
+    std::function<void(const RequestResult<InactiveRangesInfo> &)> Receiver;
+    SwiftInvocationRef Invok;
+
+  public:
+    IfConfigConsumer(
+        std::function<void(const RequestResult<InactiveRangesInfo> &)> Receiver,
+        SwiftInvocationRef Invok)
+        : Receiver(std::move(Receiver)), Invok(Invok) {}
+
+    // FIXME: Don't silently eat errors here.
+    void handlePrimaryAST(ASTUnitRef AstUnit) override {
+      auto &SrcFile = AstUnit->getPrimarySourceFile();
+      SmallVector<IfConfigInfo> Configs;
+      auto BufferID = SrcFile.getBufferID();
+      if (!BufferID)
+        return;
+      IfConfigScanner Scanner(SrcFile, *BufferID, Configs);
+      Scanner.walk(SrcFile);
+
+      // Sort by offset so nested decls are reported
+      // in source order (not tree order).
```

Comment:
I still would have thought this would be fine, what was the case you were seeing? We should be visiting source order and if that's not the case, I'd still say that's a bug. I'm fine if you just want to sort for now, but having the example would be handy :)

Reply:
The case I saw (and is covered by a test) is:
```
1. #if FOO
2.    #if BAR
3.    #endif
4. #else
5. #endif
```

The order it traverses the tree ends up with:
```
1 - #if FOO
4 - #else
2 -     #if BAR
```
So the sort simply makes that the "visual" order.

If there's a better way to traverse the tree, that could perhaps remove the need to sort. But I don't expect it to be a performance issue.

Reply:
That seems very odd to me 🤔 Definitely wouldn't have expected that. I'm not so concerned about the performance, just that the walk *should* be in source order already. Sounds like we have a bug there.